### PR TITLE
feature(upgrades): Introduces a new upgrading feature

### PR DESCRIPTION
**Pull request by juho-jaakkola**
image

### CODE REVIEW

File: views/default/js/elgg/upgrader.js [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+		}
+
+		hours = formatDigits(hours);
+		minutes = formatDigits(minutes);
+		seconds = formatDigits(seconds);
+
+		var value = hours + ':' + minutes + ':' + seconds;
+
+		timer.html(value);
+	};
```
Comment (juho-jaakkola) [inline views/default/js/elgg/upgrader.js]: This should probably go to its own AMD module. Perhaps nowadays there's even some existing library that could take care of the job. (I didn't find one back when I made the code.)

File: views/default/js/elgg/upgrader.js [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+					spinner.stop();
+					$('#upgrade-finished').removeClass('hidden');
+				}
+			}
+		};
+
+		// We use post() instead of action() so we can catch error messages
+		// and display them manually underneath the upgrade view.
+		return elgg.post('action/admin/upgrade', options);
+	};
```
Comment (juho-jaakkola) [inline views/default/js/elgg/upgrader.js]: Meh, this is really messy.

File: mod/blog/lib/upgrades.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+		"date": 2015052200,
+		"class": "Blog\\Upgrades\\FancyBlogUpgrade2015061700"
+	},
+	{
+		"title": "Another blog upgrade",
+		"description": "Lorem ipsum dolor sit amet",
+		"date": 2015052201,
+		"class": "Blog\\Upgrades\\Bar"
+	}
+]
```
Comment (juho-jaakkola) [inline mod/blog/lib/upgrades.json]: This file is currently not being used. The upgrades are instead registered in start.php using `'register', 'upgrades'` hook. See below.

File: engine/lib/admin.php [hunk 4]
```diff
@@ -226,11 +226,8 @@ function _elgg_admin_init() {
 	elgg_register_action('admin/site/set_robots', '', 'admin');
 	elgg_register_action('admin/site/set_maintenance_mode', '', 'admin');
 
-	elgg_register_action('admin/upgrades/upgrade_comments', '', 'admin');
-	elgg_register_action('admin/upgrades/upgrade_datadirs', '', 'admin');
-	elgg_register_action('admin/upgrades/upgrade_discussion_replies', '', 'admin');
-	elgg_register_action('admin/upgrades/upgrade_comments_access', '', 'admin');
```
Comment (juho-jaakkola) [inline engine/lib/admin.php]: The new way doesn't require every upgrade to have their own action and view like in Elgg 2.

File: engine/classes/Elgg/BatchUpgrader.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		$ENTITY_SHOW_HIDDEN_OVERRIDE = true;
+
+		// from engine/start.php
+		global $START_MICROTIME;
+
+		do {
+			$this->upgrade->getUpgrade()->run();
+
+			// TODO(juho) Remove after debugging
+			sleep(1);
```
Comment (mrclay) [inline engine/classes/Elgg/BatchUpgrader.php]: Delete me

File: engine/classes/Elgg/Upgrade/Locator.php [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+			$upgrade->description = "{$plugin_id}:upgrade:{$version}:description";
+			$upgrade->save();
+
+			$pending_upgrades = true;
+		}
+
+		return $pending_upgrades;
+
+		/*
+		$plugins = $this->plugins->find('all');
```
Comment (mrclay) [inline engine/classes/Elgg/Upgrade/Locator.php]: Remove

File: views/default/js/elgg/upgrader.js (hunk 2 above)
Comment (mrclay) [inline views/default/js/elgg/upgrader.js]: Mess can be cleaned up. Mark it all private.

**Comment by hypeJunction** [comment]
Feature request: allow defining dependencies on other upgrades.

**Comment by mrclay** [comment]
> dependencies on other upgrades

Strawman: Call a hook before operation can begin. It gathers statements about blocking upgrades. If no statements, allow operation.

**Comment by mrclay** [comment]
...Trying to keep it simple. Assumes our API can even query for (in)completed upgrades.

**Comment by juho-jaakkola** [comment]
I was thinking about passing in blocking upgrades like this:
```php

elgg_register_plugin_hook_handler('register', 'upgrades', function($hook, $type, $return, $params) {
	$return[] = array(
		'plugin_id' => 'blog',
		'version' => 2015061500,
		'class' => 'Blog\Upgrades\FancyBlogUpgrade2015061700',
		'deps' => [
			'blog:2016052900', // <--- These here
			'core:2016120500',
		),
	);
	return $return;
});
```

**Comment by juho-jaakkola** [comment]
But lets get this in, and implement the upgrade dependencies support later.

**Comment by mrclay** [comment]
Any specific instructions to test this? Any clean up needed before merge?

**Comment by juho-jaakkola** [comment]
Added testing instructions and some TODOs.

File: engine/classes/Elgg/Upgrade/Result.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+	private $success_count = 0;
+
+	/**
+	 * Add new error message to the batch
+	 *
+	 * @param string $message Error message
+	 * @return void
+	 */
+	public function setError($message) {
```
Comment (mrclay) [inline engine/classes/Elgg/Upgrade/Result.php]: addError

File: engine/classes/Elgg/Upgrade/Result.php [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+	public function getFailureCount() {
+		return $this->failure_count;
+	}
+
+	/**
+	 * Set an item as successfully upgraded
+	 *
+	 * @return void
+	 */
+	public function setSuccess() {
```
Comment (mrclay) [inline engine/classes/Elgg/Upgrade/Result.php]: `addSuccesses($num = 1)`?

File: engine/classes/Elgg/Upgrade/Result.php [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+	/**
+	 * Increment failure count
+	 *
+	 * This must be called every time a single item failes to
+	 * be upgraded.
+	 *
+	 * @return void
+	 */
+	public function setFailure() {
```
Comment (mrclay) [inline engine/classes/Elgg/Upgrade/Result.php]: `addFailures($num = 1)`

File: engine/classes/Elgg/Upgrade/Result.php [hunk 10]
```diff
@@ -0,0 +1,72 @@
+<?php
+
+namespace Elgg\Upgrade;
+
+/**
+ * Result of a single BatchUpgrade run
+ */
+class Result {
```
Comment (mrclay) [inline engine/classes/Elgg/Upgrade/Result.php]: I assume this gets serialized between runs? Document that and maybe mark this `final` for now.

**Comment by juho-jaakkola** [comment]
Not yet perfect, but merging anyway. It's easier to spot and fix possible shortcomings once we have some actual upgrades to test this on.

File: docs/guides/upgrading-data.rst [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+The basic structure of the class is the following:
+
+.. code:: php
+
+	<?php
+	
+	namespace Blog\Upgrades;
+	
+	use Elgg\Upgrade\Batch;
```
Comment (hypeJunction) [inline docs/guides/upgrading-data.rst]: Should be `use Elgg\Upgrade\Batch as BatchUpgrade`

**Comment by hypeJunction** [comment]
This shouldn't have been merged without proper tests. I am now spending time debugging it. Ach jo.

**Comment by juho-jaakkola** [comment]
Wasn't meant to be usable yet. Just something to work on.

**Pull request merged**